prow/gitattributes/gitattributes.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gitattributes
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/github"
+)
+
+// ghFileClient scopes to the only relevant functionality we require of a github client.
+type ghFileClient interface {
+	GetFile(org, repo, filepath, commit string) ([]byte, error)
+}
+
+// Pattern defines a single gitattributes pattern.
+type Pattern interface {
+	// Match matches the given path to the pattern.
+	Match(path string) bool
+}
+
+// Group is a logical collection of files. Check for a file's
+// inclusion in the group using the Match method.
+type Group struct {
+	LinguistGeneratedPatterns []Pattern
+}
+
+// NewGroup reads the .gitattributes file in the root of the repository only.
+func NewGroup(gc ghFileClient, owner, repo, sha string) (*Group, error) {
```

Comment:
If you really wanted to it seems like we could close over the inputs to the GetFile call, they're static and we do nothing but use them for the call

Reply:
Sorry, I don't see what you mean...

Reply:
Could have the signature

```go
func NewGroup(gitAttributesContent func() ([]byte, error)) (*Group, error) {
```

And call it like

```
NewGroup(func() ([]byte, error) { return gc.GetFile(owner, repo, gitAttributesFile, sha) })
```

If you didn't want to have to mock out the git client in your tests here

Reply:
So, cool! Indeed I will do that.